### PR TITLE
[#IOPID-328] Store extra parameter on login step

### PR DESCRIPTION
**Pull request by gquadrati**
#### List of Changes

- Update redis_cache_provider to store and retrieve extra parameters, if any
- Add new `IExtraLoginRequestParamConfig` for defining codec and requestMapper
- Add tests

#### Motivation and Context

Enable `withSpid` to store extra parameters on Redis during login step, and return it to user-defined acs callback.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/example.ts [hunk 1]
```diff
@@ -136,8 +144,11 @@ const samlConfig: SamlConfig = {
   validateInResponseTo: true,
 };
 
-const acs: AssertionConsumerServiceT = async (payload) => {
-  logger.info("acs:%s", JSON.stringify(payload));
+const acs: AssertionConsumerServiceT<ExtraParamsT> = async (
+  payload,
+  extraParams
+) => {
+  logger.info("acs:%s%s", JSON.stringify(payload), JSON.stringify(extraParams));
```
Comment (BurnedMarshal) [inline src/example.ts:151]: This log is needed or was added only to test the implementation?

File: src/index.ts [hunk 2]
```diff
@@ -79,8 +81,24 @@ export interface IEventInfo {
 
 export type EventTracker = (params: IEventInfo) => void;
 
+export interface IExtraLoginRequestParamConfig<
+  T extends Record<string, unknown>
+> {
+  // The codec to decode extra params from Redis cache provider
+  readonly codec: t.Type<T>;
+  // Extracts extra params from Login Request, that will be forwarded to the acs controller
```
Comment (BurnedMarshal) [inline src/index.ts:89]:
```suggestion
  // Extracts extra params from Login Request, that if valid will be forwarded to the acs controller
```

**Review by BurnedMarshal — APPROVED** [review]
LGTM with some stylish comments.

File: src/example.ts (hunk 1 above)
Comment (gquadrati) [reply src/example.ts:151]: It was already there, I just added a second parameter.. Anyway, it's just an example.

File: src/strategy/__tests__/saml_client.test.ts [hunk 3]
```diff
@@ -202,13 +224,61 @@ describe("CustomSamlClient#validatePostResponse", () => {
...
+      redisCacheProvider,
+      undefined,
+      expect.any(Function)
+    );
+    // Before checking the execution of the callback we must await that the TaskEither execution is completed.
+    await new Promise((resolve) => {
+      setTimeout(() => {
+        expect(mockGet).toBeCalledWith(`SAML-EXT-${expectedAuthnRequestID}`);
+        expect(mockDel).toBeCalledWith(`SAML-EXT-${expectedAuthnRequestID}`);
+        expect(mockedCallback).toBeCalledWith(
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/strategy/__tests__/saml_client.test.ts:253]: Because `extraLoginRequestParams` will be added to user even if empty.

**Pull request merged**